Site Editor Compatibility: Podcast Player block issues in editor and front end views

ramonjd opened the pull request:
This PR addresses Podcast Player block compatibility issues in the site editor, and on the frontend in block-based themes.

Fixes #19483

#### Changes proposed in this Pull Request:

1. Adds a check to see if the DOM is loaded before initializing the block on the frontend.
2. Adds a temporary workaround to inject required assets into the Site Editor iframe.

<img width="696" alt="Screen Shot 2021-04-21 at 3 13 14 pm" src="https://user-images.githubusercontent.com/6458278/115500421-96a2d980-a2b4-11eb-9679-3e9bf81b6601.png">

#### Jetpack product discussion
Related to Automattic/jetpack#19504

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
1. Using a block-based theme, e.g., TT1 add a Podcast block to the site editor. Here's an RSS feed link: https://distributed.blog/category/podcast/feed/ 😄 
2. Once loaded, you should be able to interact with the player as expected.
3. Add another one.
4. Both should work 
5. Check that, for every podcast block you add, we're not duplicating the mediaelement-css assets.
6. Take a look at the frontend. The player(s) should both load.
7. Check a few other themes for regressions.
8. Sync the changes to WPCOM and ensure that both the block and site editors work as expected.

ramonjd committed:
[not verified] For TT1 block-based theme at least, we have to wait until the dom is loaded before we execute DOM-based queries.

ramonjd committed:
[not verified] # This is a combination of 3 commits.

I don't like doing this, but there's no other way to enqueue/inject the styles we need for the podcast player into the site editor iframe.
So, using JS and the DOM, we find the CSS we need from the head of the parent window and move it into the site editor.
YAY.

Changelog

Adding null checks to the elements

we should ensure that the parent-level dom elements are there before we use them

andrewserong commented on projects/plugins/jetpack/extensions/blocks/podcast-player/utils.js:
```diff
@@ -123,3 +123,26 @@ export const getColorsObject = memoize( generateColorsObject, config => {
+		const mediaElementCSS = document.querySelector( 'link#mediaelement-css' );
+		const wpMediaElementCSS = document.querySelector( 'link#wp-mediaelement-css' );
+
+		if ( ! iframe.contentDocument.querySelector( 'link#mediaelement-css' ) ) {
```
Do we need to check that `mediaElementCSS` is not `null`, too, just in case for some reason the above querySelector on the parent document doesn't match against anything?

ramonjd replied:
We probably should do that. Thanks!

andrewserong commented on projects/plugins/jetpack/extensions/blocks/podcast-player/utils.js:
```diff
@@ -123,3 +123,26 @@ export const getColorsObject = memoize( generateColorsObject, config => {
+			mediaElementCSS.remove();
+		}
+
+		if ( ! iframe.contentDocument.querySelector( 'link#wp-mediaelement-css' ) ) {
```
Similar to above, do we need to check that `weMediaElementCSS` is not `null`, too, just in case for some reason the above querySelector on the parent document doesn't match against anything?

andrewserong reviewed (commented):
Thanks for investigating this one @ramonjd. I think this looks like a reasonable temporary workaround. I haven't tested it yet, but just wondered if we should add a couple of extra checks just in case (for some reason?) the expected CSS hasn't been enqueued in the parent document?

andrewserong commented:
Oh, it's complaining about a missing changelog entry too 🙂

ramonjd committed:
Using the block element to check whether it's sitting inside an iframe now before we load.
Abstracting the element copier to be more generic.

ramonjd committed:
Remove unused method

glendaviesnz commented:
@ramon it looks like there might be some funny race condition as it doesn't work for me locally as he useEffect never gets fired with a podcastPlayerRef.current defined.

TIL - it seems like you can't rely on ref to cause a rerender when it the dom node is assigned/changed and should use a callback ref instead. I switched it to the following locally and that seemed to fix it for me:
```js
	const podCastPlayerRef = useCallback(
		node => {
			if ( node !== null && ! hasMigratedStyles ) {
				maybeCopyElementsToSiteEditorContext(
					[ 'link#mediaelement-css', 'link#wp-mediaelement-css' ],
					node
				);
				setHasMigratedStyles( true );
			}
		},
		[ hasMigratedStyles ]
	);
```

Commit:
Switch from useEffect to ref callback for getting podcast player element ref

andrewserong commented on projects/plugins/jetpack/extensions/shared/block-editor-asset-loader.js:
```diff
@@ -0,0 +1,64 @@
+			const isElementAlreadyPresentInCurrentWindow = !! currentDoc.querySelector( selector );
+			if ( parentElementToCopy && ! isElementAlreadyPresentInCurrentWindow ) {
+				currentDoc.head.appendChild( parentElementToCopy.cloneNode() );
+				parentElementToCopy.remove();
```
Do we need to remove the element from the parent document? I don't have a practical example of this, but just imagining if for some reason a plugin were to enqueue the mediaelement for use in the sidebar inspector, which is rendered outside of the site editor iframe context?

ramonjd replied:
No, there was no pressing need. It was more of a OCD clean up manoeuvre.

You have a make a good hypothetical point though and there's no real harm in leaving it in there. I'll remove it. 👍

ramonjd replied:
I made it optional. For generic purposes, there might be a use case for removing it, maybe if the source element has styles/js that bork if their corresponding HTML isn't there or some other completely made up scenario.

andrewserong replied:
Oh, good thinking!

ramonjd committed:
Making element removal optional.

In the case of style sheets it won't matter much for our current use case (podcast player) but for other elements we might want to leave the element in both contexts. I can't for the life of me think of a reason why right now.

andrewserong commented:
This is testing well in FSE across Safari, FF, Chrome, and Edge. Unfortunately, I think this approach introduces a regression in a non-FSE theme in wpcom. Because in that context we are rendering Gutenberg within an iframe from wordpress.com, it looks like it's attempting to access the wrong parent document (e.g. the `ref` is not within the site editor iframe, but _is_ within the iframe rendered within wordpress.com, so it's now trying to break out of Gutenberg). Here's the error after entering a podcast feed URL in the placeholder:

![image](https://user-images.githubusercontent.com/14988353/115659190-8440a280-a37d-11eb-8214-2330883d1851.png)

andrewserong commented on projects/plugins/jetpack/extensions/shared/block-editor-asset-loader.js:
```diff
@@ -0,0 +1,67 @@
+ */
+export function isElementInIframe( elementRef ) {
+	const { currentWindow } = getLoadContext( elementRef );
+	return currentWindow.self !== currentWindow.top;
```
We might need to add an additional check either here or within `getLoadContext` that we're in the _correct_ iframe (the site editor) and not within another version of an iframed editor (e.g. the post/page editor iframed within `wordpress.com/post/<blog_name>/<post_id>`).

ramonjd replied:
Oh wow, thanks for double checking that. 🤕 

I did check in WPCOM when I was initially targeting `'iframe[name="editor-canvas"]'` 

I'll try to reinstate that test. 

THANK YOU

ramonjd replied:
I re-added an iframe name check. Not ideal, but the whole thing is (theoretically) transitory.

I could try to check in the back end and add a global JS variable if it might help future endeavours.

ramonjd replied:
We could also add to test that the iframe passes the same-origin-policy before we try to start injecting things into the header and bork it all again. :D

ramonjd committed:
This updates the iframe check to also look for the block editor iframe name.
It's not 100% bullet-proof, but aligns with the general theme of ephemerality in relation to these site editor bug fixes

andrewserong reviewed (dismissed):
Thanks for implementing all those changes @ramonjd! This is testing nicely for me now. I re-tested in FSE across Safari, FF, Edge, and Chrome, via wordpress.com/site-editor and directly from the the site, and double-checked that the addition of checking for the iframe name resolves the regression 👍

LGTM!

ramonjd committed:
Returning an array from maybeCopyElementsToSiteEditorContext and not mixed return types
Testing whether we indeed have access to the parent DOM before we try to avoid fatals

ramonjd commented:
>  I re-tested in FSE across Safari, FF, Edge, and Chrome, via wordpress.com/site-editor and directly from the the site

Thanks for testing so thoroughly! I added another, small condition to see if the child iframe has access to its parent (same origin). Hopefully will prevent borks if the editor iframe name ever changes.

andrewserong reviewed (approved):
Thanks for adding the additional check @ramonjd 👍 that looks safer. Just gave this another quick smoke test in the post/page editor (via wordpress.com iframe and via wp-admin) and the site editor (again via wordpress.com/site-editor and the direct /wp-admin url). LGTM! 🎉

jeherve reviewed (approved):
This tests well for me. 👍

ramonjd commented:
r224746-wpcom

ramonjd commented:
An api is being added to core Gutenberg that we may be able to use instead of the CSS shuffling: WordPress/gutenberg#31873
